### PR TITLE
[2/N] Add flag to control which rank should perform NaN check

### PR DESCRIPTION
**Pull request by kwen2501**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #134707
* #134701
* #134357
* __->__ #134345

Fixes https://github.com/pytorch/pytorch/issues/134062.
For example, in case of broadcast / scatter, only the root rank should perform the NaN check.

cc @XilunWu @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -3097,7 +3100,9 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::pointToPoint(
   // is gpuGuard needed for the if block below, or can i swap them
   at::cuda::OptionalCUDAGuard gpuGuard;
 
-  if (enableNanCheck_) {
+  // Only check for NaN for send ops, for recv ops `tensor` can be a random
+  // placeholder
+  if (enableNanCheck_ && opType == OpType::SEND) {
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:3105]: double checking- does this check work correctly for coalesced p2p too? would this function be called when creating the individual isend ops and check their nans at that time?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 2]
```diff
@@ -3455,14 +3462,16 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::broadcast(
   // avoidRecordStreams_ note: collective() will stash tensors.
   bool avoidRecordStreams = avoidRecordStreams_ || (!opts.asyncOp);
 
+  const auto root = opts.rootRank + opts.rootTensor;
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:3465]: what is opts.rootTensor?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp [hunk 3]
```diff
@@ -728,7 +728,8 @@ class TORCH_API ProcessGroupNCCL : public Backend {
       Fn fn,
       OpType opType,
       const char* profilingTitle = nullptr,
-      bool avoidRecordStreams = false);
+      bool avoidRecordStreams = false,
+      bool nanCheck = true);
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:732]: so we default to true which makes sense, bc most collectives (allreduce, allgather, gather, a2a) will need to check all inputs.  The ones that don't are special-cased (broadcast, scatter).  i think we covered them all.  
 
Is `_broadcast_oop` covered by broadcast?

**Comment by wconstab** [comment]
this one LGTM. just add tests.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 1 above)
Comment (kwen2501) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:3105]: Yeah, I think individual ops in batch P2P would call into this function too, with OpType::Send or OpType::Recv, so they should be covered.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp (hunk 3 above)
Comment (kwen2501) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:732]: Thanks for checking. Added ` _broadcast_oop` now.

**Comment by kwen2501** [comment]
> this one LGTM. just add tests.

Added a test doing P2P where NaN number exists in receive buffer -- NaN checker should not bark.

**Review by shuqiangzhang — APPROVED** [review]

**Review by wconstab — APPROVED** [review]

**Pull request closed without merging**